### ISSUE TITLE
[BUG] Async Blob Storage - Max page size in search options is not being used correctly

### ISSUE THREAD

**Issue by technical-debt-collector** [Azure/azure-sdk-for-java#33463]
**Describe the bug**
We've got a blocking API that uses the SDK to fetch and upload blobs.
I've spent the last couple of days rewriting it to async. However, the performance just seems consistently worse no matter how much logic we strip off the API in attempts to isolate the issue(s).
Additionally, the application consumes **way** more CPU than its blocking counterpart.

Profiling a smaller load test gives off the impression of Jackson being the main culprit. But, this is also after having stripped away as much as possible from the API temporarily to reduce noise while debugging, so it's only natural that Jackson and Reactor account for the majority of the samples.

***Code Snippet***
(Code is simplified and has a few steps omitted but this is the exact code we've run with the attached CPU profile)

Using JAX-RS, our API method looks like this:

```
@GET
    public Multi<TaggedBlobItem> search(@PathParam("tenant") String tenant, @Valid @BeanParam ArchiveSearchParameters searchParameters) {
        return Multi.createFrom().publisher(blobStore.searchBlobs(tenant, searchParameters));
    }

```

And our service logic looks like this:

```
public Flux<TaggedBlobItem> searchBlobs(Tenants tenant, ArchiveSearchParameters search) {
        var searchOptions = new FindBlobsOptions("<query that is omitted in this example>")
            .setMaxResultsPerPage(search.pageSize);
        return blobServiceAsyncClient
            .findBlobsByTags(searchOptions)
            .take(searchOptions.getMaxResultsPerPage(), true);
    }
```

**Expected behavior**
Equal or improved performance compared to the blocking version

**Screenshots**
Attached a screenshot of a flamegraph from when running a wrk load test using async.
```
wrk
  4 threads and 120 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.65s   240.10ms   2.00s    64.75%
    Req/Sec     7.19      5.92    40.00     79.86%
  1034 requests in 1.00m, 3.26MB read
  Socket errors: connect 0, read 0, write 0, timeout 895
Requests/sec:     17.21
Transfer/sec:     55.54KB

```
![async_profile](https://user-images.githubusercontent.com/79846697/218670375-3972e9b7-0fd9-45c7-81f1-0bd9d6adbfbd.png)

**Setup (please complete the following information):**
 - OS: Linux
 - IDE: Intellij
 - Library/Libraries: com.azure:azure-storage-blob:12.20.2 (supplied using BOM com.azure:azure-sdk-bom:1.2.9)
 - Java version: 17
 - Frameworks: Quarkus

If you suspect a dependency version mismatch (e.g. you see `NoClassDefFoundError`, `NoSuchMethodError` or similar), please check out [Troubleshoot dependency version conflict article](https://aka.ms/azsdk/java/dependency/troubleshoot) first. If it doesn't provide solution for the problem, please provide:
- verbose dependency tree (`mvn dependency:tree -Dverbose`)
- exception message, full stack trace, and any available logs

**Additional context**
Our API is a Quarkus-based application in which we make Mutiny (the Quarkus-preferred Reactive library) subscribe to the Project Reactor publishers with little else going on.

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [ ] Repro Steps Added
- [x] Setup information Added

**Comment by alzimmermsft** [comment]
Thanks for filing this @technical-debt-collector, could you also include a codesnippet of your synchronous implementation for a side-by-side comparison.

**Comment by technical-debt-collector** [comment]
To make them more comparable I've omitted equally as much from this one as we did on the async stuff and run the same wrk test again with the same specs as above.

One of the omissions was some metadata fetching for each of the returned blobs (this search returns the first 25 blobs matching the query). The vast difference in wrk results **really** makes me suspect I've done something wrong in the async one.

wrk
  4 threads and 120 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    86.71ms  108.20ms   1.81s    90.88%
    Req/Sec   439.16    137.89     1.04k    79.10%
  104509 requests in 1.00m, 311.36MB read
  Socket errors: connect 0, read 0, write 0, timeout 60
Requests/sec:   1740.84
Transfer/sec:      5.19MB


```
 @GET
    public List<TaggedBlobItem> search(@PathParam("tenant") String tenant, @Valid @BeanParam ArchiveSearchParameters searchParameters) throws IOException {
        var theTenant = verifyTenant(tenant);
        return blobStore.searchBlobs(theTenant, searchParameters);
    }
```

```
    public List<TaggedBlobItem> searchBlobs(Tenants tenant, ArchiveSearchParameters search) throws IOException {
        List<TaggedBlobItem> result = List.of();
        var searchOptions = new FindBlobsOptions("<some magic query stuff>");
        searchOptions.setMaxResultsPerPage(25);
        var pages = service.findBlobsByTags(searchOptions, null, Context.NONE).iterableByPage().iterator();
        if (pages.hasNext()) {
            try (var pagedResponse = pages.next()) {
                result = pagedResponse.getValue().stream().limit(search.pageSize).collect(Collectors.toList());
            }
        }
        return result;
    }
```

**Comment by alzimmermsft** [comment]
@technical-debt-collector it's possible that the asynchronous implementation may have been retrieving more than one find/search blobs paged response which was resulting in the difference in performance. At this time, the synchronous APIs in `BlobServiceClient` just call into `BlobServiceAsyncClient` and block the asynchronous request, so there really shouldn't be any performance differences between synchronous and asynchronous.

**Comment by technical-debt-collector** [comment]
Thanks for pointing that out! You're right, the async implementation had limited the upstream to a page size worth of elements, which is not correct. 
I've just changed this and tested some more and while it does seem to improve some we're still far off from any acceptable result. I'll come back to you asap once I find out more

**Comment by technical-debt-collector** [comment]
So there were two issues with the async implementation on my end.

```
public Flux<PagedResponse<TaggedBlobItem>> searchBlobs(Tenants tenant, ArchiveSearchParameters search) {
    var searchOptions = new FindBlobsOptions("<query>").setMaxResultsPerPage(search.pageSize);
    return service
        .findBlobsByTags(searchOptions)
        .byPage(search.pageSize)
        .take(1, true); 
}
```

1. byPage(int preferredPageSize) actually constrains the pages down to the desired size. However, setting the max results per page in the searchOptions-object doesn't seem to do anything.
2. take(long n, boolean limitRequest) should obviously not specify the page size, but one, as we in this particular case is only interested in retrieving the first page of whatever we're searching for.

The javadoc for the synchronous version mentions more about the search options. 
![image](https://user-images.githubusercontent.com/79846697/219585217-e39a0fd8-362d-446f-9ffa-a7b0c91b4342.png)
I did some quick tests just now and setting a value here does seem to work, at least when performing the search using .iterableByPage():
`service.findBlobsByTags(searchOptions, null, Context.NONE).iterableByPage().iterator();`

But why is the pagesize, as specified through FindBlobOptions.setMaxResultsPerPage(), not being used in the async implementation?
The value we set here doesn't seem to matter at all, and except for async having less javadoc about search options I can't find anything about it.

**Comment by technical-debt-collector** [comment]
Just to add, omitting the parameter in the async byPage()-chain commented above in still returns way more results than desired, so it doesn't seem to be a question about whether we're paging the results or not.

**Comment by ibrahimrabab** [comment]
Hi @technical-debt-collector 
Thanks for reaching out with this issue. I tried to test this bug locally but was unable to reproduce the error of getting more results than specified. 

Can you please provide a full code snippet for the async client and how you are reproducing this? Thanks!

**Comment by ibrahimrabab** [comment]
Hi @technical-debt-collector 
Thank you so much for providing a thorough code sample! We were able to debug and find the issue. The problem was that we gave precedence to the `.byPage()`, so if `.byPage()` was called with results=null, the max results was also set to null even if max results was specified in the `FindBlobsOptions`. 
We've gone ahead and fixed this, now when `.byPage()` is called, it will check if maxResults is set from `FindBlobsOptions`. If so, we will use that value, and if not, all results will be returned. This change should be part of our future release cycle.
